### DISCUSSION TITLE
loss resets after resuming pretraining

### DISCUSSION THREAD

**Discussion by Onyoursix** [explosion/spaCy#9080] [Help: Coding & Implementations]
I've been pretraining on a fairly large data set for the past 24 hours and got kicked off my GPU session. I resumed my pre-training per the docs but it seems that the loss has basically reset. I used the -r /path/to/latest/model.bin with -er as the next epoch to resume on.

here is what my loss looks like graphically from the log.jsonl file, is this normal?

![Screenshot from 2021-08-27 23-40-51](https://user-images.githubusercontent.com/32439975/131207665-6c98c17e-df63-44fe-9aba-185d16e83d47.png)

**Comment by polm** [comment]
Confirmed this is happening using the [sample textcat project](https://github.com/explosion/projects/tree/v3/tutorials/textcat_docs_issues). 

I don't think it's intended behavior, but it might not be a problem, I'll need to look at it more.

**Reply by Onyoursix** [reply]
Thanks for the reply polm. I tried to run this again from scratch to see what would happen over the course of three days. For context I'm running this on a GPU on Google Colab so at the 24hr mark I'm getting booted off my GPU session. The 2nd restart it took almost a full 24 hours to just make it back down to where the loss was previously. 

Here is the entire graph
![Screenshot from 2021-08-30 11-08-00](https://user-images.githubusercontent.com/32439975/131377949-74b39144-e443-4c5e-b453-7f82559d9390.png)

Last 50 epochs
![Screenshot from 2021-08-30 11-12-11](https://user-images.githubusercontent.com/32439975/131377973-0d61efb3-f27a-4478-afe4-16dbcfc4f80c.png)

Only two epochs made it past the previous point before disconnect.
![Screenshot from 2021-08-30 11-09-58](https://user-images.githubusercontent.com/32439975/131378085-20506582-cc3a-432b-8789-49efe1ad21ea.png)

The gains are not significant, but they are still gains, I'm pretraining on just under 1M examples. I don't know if any of that is helpful to you at all. Thanks for all you and your team do!

**Reply by polm** [reply]
Thanks for the extra data.

Basically this is happening because we serialize the model, but not the optimizer (Adam/SGD/etc). There are a couple of problems this can cause:

1. If you have a learning rate or other schedule, it will be reset. This could be very bad, but it doesn't seem to be affecting you directly. We can fix this pretty easily by stepping the schedulers without doing computation.
2. Optimizers can (or always do?) have state, for example Adam tracks momentum. We can't get this back without doing calculations so we either have to serialize the optimizer or give up. I *think* these values should generally go back to the right place even if not saved, though your example indicates that while that's happening it's not as fast as could be hoped.

It sounds like we'll probably have to serialize the optimizer to proper training resume working. I don't think that'll be that complicated but it'll probably mean serializing an extra file so we have to think a bit about how to do it. In the meantime, resuming will work but you'll get discontinuous loss like this and it'll take a little while to catch up.

Sorry for the trouble but we'll work on fixing it and post here when it's done.

**Reply by Onyoursix** [reply]
Thanks for the info polm. No worries, I'll keep my eye on this thread and watch for the update. I appreciate you explaining all of this and looking into the issue.
